### PR TITLE
BUG: Fix overflow warning on apple silicon

### PR DESCRIPTION
**Pull request by isuruf**
Alternative to https://github.com/numpy/numpy/pull/17712

### CODE REVIEW

**Comment by seberg** [comment]
@isuruf, I know that did not work in gh-18555 which may actually just be the identical issue?

Before we add more clang workaround, can you make sure to compiling with `-ffp-exception-behavior=strict`? We have to make that flag default on clang and make sure to not add any workarounds for clang bugs that are actually just bad compiler defaults.

I am not on top of these efforts here, so ignore if this is obviously unhelpful...

**Comment by isuruf** [comment]
> `-ffp-exception-behavior=strict`

The exception is still there if I add that flag without this PR.

**Comment by rgommers** [comment]
The one PyPy failure seems unrelated:
```
____________________ TestRandomDist.test_choice_exceptions _____________________
...
TypeError: __format__() missing 1 required positional argument: 'spec'
```
Let's see if there's more opinions on a preference of this PR vs. gh-17712. I'd be inclined to merge this one.

**Comment by erykoff** [comment]
I think this would be fine.  I think my solution on #17712 would be a bit cleaner (and have the advantage of unifying the code path of single-value and vector computations), but I was stymied by the time taken to make it work with the 128-bit code path.  This is certainly a simpler and more direct solution.

File: numpy/core/src/umath/loops.c.src [hunk 1]
```diff
@@ -2526,6 +2526,21 @@ NPY_NO_EXPORT void
         const @ftype@ in1i = ((@ftype@ *)ip1)[1];
         const @ftype@ in2r = ((@ftype@ *)ip2)[0];
         const @ftype@ in2i = ((@ftype@ *)ip2)[1];
+#if defined(__APPLE__) && defined(__aarch64__)
+        // On macos-arm64, if this block of code was not there,
```
Comment (mattip) [inline numpy/core/src/umath/loops.c.src:2530]:
```suggestion
        // On macos-arm64 without this block of code,
```

**Pull request merged**